File: DesignPatternDemo/Pattern/Decoration/Decorator.m
//
//  Decorator.m
//  DesignPatternDemo
//
//  Created by ruantong on 2019/5/11.
//  Copyright © 2019 wupeng. All rights reserved.
//

#import "Decorator.h"

@interface Decorator()
{
    SchoolReport * _sr;
}
@end

@implementation Decorator

- (instancetype)initWithSr:(SchoolReport *)sr{
    self = [super init];
    if (self) {
        _sr = sr;
    }
    return self;
}

- (void)report{
    [_sr report];
}

- (void)sign:(NSString *)name{
    [_sr sign:name];
}

@end

File: DesignPatternDemo/Pattern/Decoration/Decorator.swift
//
//  Decorator.swift
//  DesignPatternDemo
//
//  Created by wupeng on 2020/3/4.
//  Copyright © 2020 wupeng. All rights reserved.
//

import Foundation

class Decorator: SchoolReport {
    var sr:SchoolReport
    
    public init(sr:SchoolReport) {
        self.sr = sr
    }
    
    override func report() {
        self.sr.report()
    }
    
    override func sign(name: NSString) {
        self.sr.sign(name: name)
    }
}

class HighScoreDecorator: Decorator {
    
    override func report() {
        super.report()
        self.reportHighScore()
    }
    
    func reportHighScore(){
        print("这次考试语文最高是75，数学78，自然是80")
    }
    
    override func sign(name: NSString) {
        super.sign(name: name)
        print("我是高分修饰签名")
    }
}

class SortDecorator: Decorator {

    override func report() {
        super.report()
        self.reportSort()
    }
    
    func reportSort(){
        print("我的排名第38名")
    }
    
    override func sign(name: NSString) {
        super.sign(name: name)
        print("我是排名修饰签名")
    }
}
